### DISCUSSION TITLE
KEM Question about restoring encapsulated secret by decapsulation

### DISCUSSION THREAD

**Discussion by sungmin-net** [open-quantum-safe/liboqs#1380] [Q&A] [answered]
Dear OQS team,
I'm newbie in OQS then please understand my questions.

I tried the sample KEM code in "https://openquantumsafe.org/liboqs/examples/kem.html".
In my understanding, the variable 'shared_secret_d' should have same value with the variable 'shared_secret_e' after OQS_KEM_frodokem_640_aes_decaps() because the variable 'ciphtertext' is the encapsulated value for 'shared_secret_d'.

In my trial, I gave 0xC0, 0xC1, 0xC2, ... , 0xCF to 'shared_secret_**e**' and printed 'shared_secret_**d**' after OQS_KEM_frodokem_640_aes_decaps(). I expected that 'shared_secret_**d**' has 0xC0, 0xC1, 0xC2, ... , 0xCF but printing 'shared_secret_**d**' showed 0xCA, 0xB9, 0x43, 0x43, 0x32, 0x28, 0x29, 0xEB, 0x81 , 0x94, 0x73, 0xB2, 0x1C, 0x1B, 0x0E, 0x2B.

Would you please let me know how I can restore original value 0xC0, 0xC1, 0xC2, ... , 0xCF after decapsulation?

Thanks in advance,
Sungmin.

**Comment by baentsch** [comment, marked as answer]
Can you please share the complete code you're running? Looking at this statement

> In my trial, I gave 0xC0, 0xC1, 0xC2, ... , 0xCF to 'shared_secret_e' and printed 'shared_secret_d' after OQS_KEM_frodokem_640_aes_decaps(). 

it seems you are not using encaps/decaps as a pair (as it needs to be). See a [complete example highlighting  identity test here](https://github.com/open-quantum-safe/liboqs/blob/423b720c38985fd3fe8cc4874806fe0c5be6266e/tests/test_kem.c#L120).

**Reply by sungmin-net** [reply]
Thank you very much @baentsch 
I misunderstood that shared_secret_e is an input value for encaps() API to produce ciphertext. 
Now I understand that shared_secret_e is output value from encaps() API and decaps() makes shared_secret_d to be same with shared_secret_e using the ciphertext.
Thanks again!
